src/VisualStudio/Core/Def/Telemetry/AggregatingTelemetryLog.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.VisualStudio.Telemetry;
+using Microsoft.VisualStudio.Telemetry.Metrics;
+using Microsoft.VisualStudio.Telemetry.Metrics.Events;
+
+namespace Microsoft.CodeAnalysis.Telemetry
+{
+    /// <summary>
+    /// Provides a wrapper around the VSTelemetry histogram APIs to support aggregated telemetry. Each instance
+    /// of this class corresponds to a specific FunctionId operation and can support aggregated values for each
+    /// metric name logged.
+    /// </summary>
+    internal sealed class AggregatingTelemetryLog : ITelemetryLog
+    {
+        private readonly IMeter _meter;
+        private readonly TelemetrySession _session;
+        private readonly Dictionary<string, IHistogram<int>> _histograms;
+        private readonly object _lock;
+        private readonly HistogramConfiguration? _histogramConfiguration;
+        private readonly string _eventName;
+
+        private const string MeterVersion = "0.39";
+
+        public AggregatingTelemetryLog(TelemetrySession session, FunctionId functionId, double[]? bucketBoundaries = null)
+        {
+            var meterName = TelemetryLogger.GetPropertyName(functionId, "meter");
+            var meterProvider = new VSTelemetryMeterProvider();
+
+            _session = session;
+            _meter = meterProvider.CreateMeter(meterName, version: MeterVersion);
+            _histograms = new();
+            _lock = new();
+            _eventName = TelemetryLogger.GetEventName(functionId);
+
+            if (bucketBoundaries != null)
+            {
+                _histogramConfiguration = new HistogramConfiguration(bucketBoundaries);
+            }
+        }
+
+        /// <summary>
+        /// Adds aggregated information for the metric and value passed in via logMessage. KeyValueLogMessage.Properties
+        /// is searched for appropriate entries to determine the metric name and value.
+        /// </summary>
+        /// <param name="logMessage"></param>
+        public void Log(LogMessage logMessage)
+        {
+            if (!IsEnabled)
+                return;
+
+            if (logMessage is not KeyValueLogMessage kvLogMessage)
+                return;
+
+            string? metricName = null;
+            int? value = null;
+            foreach (var kvp in kvLogMessage.Properties)
+            {
+                switch (kvp.Value)
+                {
+                    case string s:
+                        metricName = s;
```

Comment:
So the expectation is there will be only a single string valued property and a single int valued property in the message? Is this asserted somewhere? Otherwise, this might lead to a race condition where we end up using different values based on the order in which properties are iterated.
Can we add asserts in both the switch cases here that corresponding local is null prior to the assignment?

Reply:
A more hardened approach might be to have explicit property names that are used here for both the metric name and value. I see that the Timed block uses `Name` and `Delay` as the property names - if that is the only place creating log messages which reach here, we can consider looking for these specific property values instead of iterating all the properties.

Reply:
I didn't go this approach originally as it required both sides to know the names being used, but in hindsight that's probably cleaner than what I presented here. Changed in commit 3.